File: server/app/model/schema/question/question.ts
/* eslint-disable no-underscore-dangle */ // Attributs de Mongo utilisent un underscore
import { QuestionChoice } from '@app/model/schema/question-choice/question-choice';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Document, Types } from 'mongoose';

export type QuestionDocument = Question & Document;

@Schema({
    toJSON: {
        virtuals: true,
        versionKey: false,
        transform: (doc, ret) => {
            ret.id = ret._id.toHexString();
            delete ret._id;
            delete ret.__v;

            if (ret.choices) {
                ret.choices.forEach((choice) => {
                    delete choice._id;
                    delete choice.id;
                });
            }
        },
    },
    toObject: { virtuals: true },
})
export class Question {
    @ApiProperty()
    @Prop({ required: true })
    type: string;

    @ApiProperty()
    @Prop({ required: true })
    text: string;

    @ApiProperty()
    @Prop({ required: true })
    points: number;

    @ApiProperty()
    @Prop({
        type: [
            {
                text: String,
                isCorrect: { type: Boolean, default: false },
            },
        ],
        default: [],
    })
    choices: Types.Array<QuestionChoice>;

    @ApiProperty({ description: 'The last modification date of the question' })
    @Prop({ required: false, default: new Date().toString() })
    lastModified: string;
}

export const questionSchema = SchemaFactory.createForClass(Question);

File: client_leger/lib/models/question.dart
import 'qre_attributes.dart';
import 'question_choice.dart';

class Question {
  String? id;
  String type;
  String text;
  String? image;
  int points;
  List<QuestionChoice>? choices;
  String? lastModified;
  QreAttributes? qreAttributes;

  Question({
    this.id,
    required this.type,
    required this.text,
    required this.points,
    this.image,
    this.choices,
    this.lastModified,
    this.qreAttributes,
  });

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      id: json['id'],
      type: json['type'],
      text: json['text'],
      points: json['points'],
      choices: json['choices'] != null
          ? (json['choices'] as List)
              .map((choice) => QuestionChoice.fromJson(choice))
              .toList()
          : null,
      image: json['image'],
      lastModified: json['lastModified'],
      qreAttributes: json['qreAttributes'] != null
          ? QreAttributes.fromJson(json['qreAttributes'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'text': text,
      'image': image,
      'points': points,
      'choices': choices?.map((choice) => choice.toJson()).toList(),
      'lastModified': lastModified,
      'qreAttributes': qreAttributes?.toJson(),
    };
  }
}
